// react-native-full-screen-notification-incoming-call/lib/module/index.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { DeviceEventEmitter, NativeEventEmitter, NativeModules, Platform } from 'react-native';
const isAndroid = Platform.OS === 'android';
const RNNotificationIncomingCall = NativeModules.FullScreenNotificationIncomingCall;
let eventEmitter;

if (isAndroid) {
  eventEmitter = new NativeEventEmitter(RNNotificationIncomingCall);
}

var RNNotificationEvent;

(function (RNNotificationEvent) {
  RNNotificationEvent["RNNotificationAnswerAction"] = "RNNotificationAnswerAction";
  RNNotificationEvent["RNNotificationEndCallAction"] = "RNNotificationEndCallAction";
})(RNNotificationEvent || (RNNotificationEvent = {}));

var CallAction;

(function (CallAction) {
  CallAction["ACTION_END_CALL"] = "ACTION_END_CALL";
  CallAction["ACTION_REJECTED_CALL"] = "ACTION_REJECTED_CALL";
  CallAction["ACTION_HIDE_CALL"] = "ACTION_HIDE_CALL";
  CallAction["ACTION_SHOW_INCOMING_CALL"] = "ACTION_SHOW_INCOMING_CALL";
  CallAction["HIDE_NOTIFICATION_INCOMING_CALL"] = "HIDE_NOTIFICATION_INCOMING_CALL";
  CallAction["ACTION_PRESS_ANSWER_CALL"] = "ACTION_PRESS_ANSWER_CALL";
  CallAction["ACTION_PRESS_DECLINE_CALL"] = "ACTION_PRESS_DECLINE_CALL";
  CallAction["ACTION_START_ACTIVITY"] = "ACTION_START_ACTIVITY";
})(CallAction || (CallAction = {}));

class RNNotificationCall {
  constructor() {
    _defineProperty(this, "_notificationEventHandlers", void 0);

    _defineProperty(this, "displayNotification", (uuid, avatar, timeout, foregroundOptions) => {
      if (!isAndroid) return;
      RNNotificationIncomingCall.displayNotification(uuid, avatar, timeout ? timeout : 0, foregroundOptions);
    });

    _defineProperty(this, "hideNotification", () => {
      if (!isAndroid) return;
      RNNotificationIncomingCall.hideNotification();
    });

    _defineProperty(this, "backToApp", () => {
      if (!isAndroid) return;
      RNNotificationIncomingCall.backToApp();
    });

    _defineProperty(this, "addEventListener", (type, handler) => {
      if (!isAndroid) return;
      let listener;

      if (type === 'answer') {
        listener = eventEmitter.addListener(RNNotificationEvent.RNNotificationAnswerAction, eventPayload => {
          handler(eventPayload);
        });
      } else if (type === 'endCall') {
        listener = eventEmitter.addListener(RNNotificationEvent.RNNotificationEndCallAction, eventPayload => {
          handler(eventPayload);
        });
      } else {
        return;
      }

      this._notificationEventHandlers.set(type, listener);
    });

    _defineProperty(this, "removeEventListener", type => {
      if (!isAndroid) return;

      const listener = this._notificationEventHandlers.get(type);

      if (!listener) {
        return;
      }

      listener.remove();

      this._notificationEventHandlers.delete(type);
    });

    _defineProperty(this, "declineCall", (uuid, payload) => {
      this.hideNotification();
      const data = {
        callUUID: uuid,
        endAction: CallAction.ACTION_REJECTED_CALL,
        payload
      };
      DeviceEventEmitter.emit(RNNotificationEvent.RNNotificationEndCallAction, data);
    });

    _defineProperty(this, "answerCall", (uuid, payload) => {
      this.hideNotification();
      const data = {
        callUUID: uuid,
        payload
      };
      DeviceEventEmitter.emit(RNNotificationEvent.RNNotificationAnswerAction, data);
    });

    this._notificationEventHandlers = new Map();
  }

}

export default new RNNotificationCall();
